[docs] Add "Migration" section to sidebar and revise v4-v5 content (#32740)
mui/material-ui: docs/data/material/migration/migration-v4/migrating-from-jss.md

# Migrating from JSS (optional)

<p class="description">This guide explains how to migrate from JSS to Emotion when updating from Material UI v4 to v5.</p>

## Material UI v5 migration

1. [Getting started](/material-ui/migration/migration-v4/)
2. [Breaking changes part one: style and theme](/material-ui/migration/v5-style-changes/)
3. [Breaking changes part two: components](/material-ui/migration/v5-component-changes/)
4. Migrating from JSS 👈 _you are here_
5. [Troubleshooting](/material-ui/migration/troubleshooting/)

## Migrating from JSS to Emotion

> One of the biggest changes in v5 is the replacement of JSS for [Emotion](https://emotion.sh/docs/introduction) (or [styled-components](https://styled-components.com/) as an alternative) as a default styling solution .

Reviewer: ```suggestion
One of the biggest changes in v5 is the promotion of [Emotion](https://emotion.sh/docs/introduction) (or [styled-components](https://styled-components.com/) as an alternative) as a default styling solution.
```

+  root: `${PREFIX}-root`,
+  cta: `${PREFIX}-cta`,
+  content: `${PREFIX}-content`,
+}
+const Root = styled('div')(({ theme }) => ({
+  [`&.${classes.root}`]: {
+    display: 'flex',
+    alignItems: 'center',
+    backgroundColor: theme.palette.primary.main
+  },
+  [`& .${classes.cta}`]: {
+    borderRadius: theme.shape.radius
+  },
+  [`& .${classes.content}`]: {
+    color: theme.palette.common.white,
+    fontSize: 16,
+    lineHeight: 1.7
+  },
+}))

 export const MyCard = () => {
-  const classes = useStyles();
   return (
-    <div className={classes.root}>
+    <Root className={classes.root}>
       {/* The benefit of this approach is that the code inside Root stays the same. */}
       <Typography className={classes.content}>...</Typography>
       <Button className={classes.cta}>Go</Button>
-    </div>
+    </Root>
   )
 }
```

:::warning
You should run this codemod on a small chunk of files and then check the changes before continuing, because in some cases you might need to adjust the code after the transformation—this codemod won't cover all cases.
:::

#### Manual

We recommend `sx` API over `styled` for creating responsive styles or overriding minor CSS.

> [Read more about `sx` here](/system/the-sx-prop/#main-content).

Reviewer: Oops

```suggestion
[Read more about `sx` here](/system/the-sx-prop/).
```

#### withStyles()

`tss-react` also features a [type-safe implementation](https://docs.tss-react.dev/page-1/withstyles) of [v4's `withStyles()`](https://v4.mui.com/styles/api/#withstyles-styles-options-higher-order-component).

:::info
The equivalent of the `$` syntax is also supported in tss's `withStyles()`.
[See doc](https://docs.tss-react.dev/nested-selectors#withstyles).
:::

```diff
-import Button from '@material-ui/core/Button';
+import Button from '@mui/material/Button';
-import withStyles from '@material-ui/styles/withStyles';
+import { withStyles } from 'tss-react/mui';

 const MyCustomButton = withStyles(
+  Button,
   (theme) => ({
     root: {
       minHeight: '30px',
     },
     textPrimary: {
       color: theme.palette.text.primary,
     },
     '@media (min-width: 960px)': {
       textPrimary: {
         fontWeight: 'bold',
       },
     },
   }),
-)(Button);
+);

 export default MyCustomButton;
```

#### Theme style overrides

[Global theme overrides](https://v4.mui.com/customization/components/#global-theme-override) are supported out of the box by TSS.

> Follow the instructions in the relevant section of the [Breaking changes](/material-ui/migration/v5-style-changes#restructure-component-definitions) doc, and [provide a `name` to `makeStyles`](https://docs.tss-react.dev/page-1/makestyles-usestyles#naming-the-stylesheets-useful-for-debugging-and-theme-style-overrides).

Reviewer 2: ```suggestion
Follow the instructions in the relevant section of the [Breaking changes](/material-ui/migration/v5-style-changes#restructure-component-definitions) doc, and [provide a `name` to `makeStyles`](https://docs.tss-react.dev/page-1/makestyles-usestyles/#naming-the-stylesheets-useful-for-debugging-and-theme-style-overrides).
```

Reviewer 2: This is breaking the build

# Migrating from JSS

<p class="description">This guide explains how to migrate from JSS to Emotion when updating from Material UI v4 to v5.</p>

## Material UI v5 migration

1. [Getting started](/material-ui/migration/migration-v4/)
2. [Breaking changes part one: style and theme](/material-ui/migration/v5-style-changes/)
3. [Breaking changes part two: components](/material-ui/migration/v5-component-changes/)

> 4. Migrating from JSS <— _you are here_

Reviewer 3: Agree with the emoji

5. [Troubleshooting](/material-ui/migration/troubleshooting/)

## Migrating from JSS to Emotion

One of the biggest changes in v5 is the swapping of JSS for [Emotion](https://emotion.sh/docs/introduction) (or styled-components as an alternative).

Note that you may continue to use JSS after migrating to v5, and when you're ready to move over to the new styling engine, you can refactor your components progressively.

> This document reviews all the steps necessary to migrate away from JSS.

Reviewer 2: It's important here to describe which steps from here are a must (e.g. you need to migrate the `styleOverrides` as they are processed with emotion), but you don't need to do the rest if you don't want to (or at least you can do it progressively).

Considering this, maybe we should move the section "Migrate theme styleOverrides to Emotion" to https://deploy-preview-32740--material-ui.netlify.app/material-ui/migration/v5-style-changes/ and than mark this page as (optional) in both the sidebar and the title of the page. I think this will help people better understand what is mandatory and what is optional step in the migration.
